### DISCUSSION TITLE
How to troubleshoot @apply circular dependency error when the code is different to what the error suggests

### DISCUSSION THREAD

**Discussion by AccessEleanorJohnson** [tailwindlabs/tailwindcss#12392] [Help]
I'm getting the following error:
`You cannot '@apply' the 'text-center' utility here because it creates a circular dependency.`

However, the section of SASS that it points to is :
```
.brand-anchor {
    @apply relative;
    @apply pl-[0.6em];
    @apply pt-[0.48em];

    .text-center & {
      @apply pl-0;
    }

    &::before {
      @apply absolute;
      @apply inline-block;
      @apply -translate-x-[0.6em];
      @apply -translate-y-[0.48em];
      @apply font-fontello;
      @apply content-["\e82d"];
      @apply font-thin;
    }
  }
```

Text-center is not being applied, it is part of the selector that a different utility is being applied to (`@apply pl-0`). Obviously there is something happening here that Tailwind is not happy with, but this error message is not helpful to me. How can I troubleshoot this in order to find out what the real problem is?

From trying to look this up, usually someone would be trying the `@apply text-center` to a class of the same name, but in my case it is not. `@apply text-center` is used further up the file, but is is separate from these styles and is being applied to the selector `.align-center`.

**Comment by wongjn** [comment]
I believe the error could be coming from this rule:

```scss
.text-center & {
  @apply pl-0;
}
```
It could be that some class name resolution later on creates the circular dependency.

To work around the problem, consider relying less on `@apply` and use `theme()` where needed. For example, if you do have `@apply text-center` somewhere, consider using `text-align: center` instead.

Adam Wathan, creator of Tailwind does advocate avoiding `@apply`:
https://twitter.com/adamwathan/status/1226511611592085504
https://twitter.com/adamwathan/status/1559250403547652097

**Reply by AccessEleanorJohnson** [reply]
Yes, I had already figured it was picking up that selector. When I remove all uses of `@apply text-center` the error goes away, but I am still none the wiser about the actual issue. None of the other uses of `text-center` relate to this selector, it feels like it is just seeing the class name is used, sees the `@apply`, then throwing an error regardless.
